lib/ipam/ipam.go

```diff
@@ -299,10 +301,43 @@ func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, a
 		logCtx = logCtx.WithField("handle", *handleID)
 	}
 	logCtx.Info("Looking up existing affinities for host")
-	affBlocks, err := c.blockReaderWriter.getAffineBlocks(ctx, host, version, pools)
+	affBlocks, affBlocksToRelease, err := c.blockReaderWriter.getAffineBlocks(ctx, host, version, pools)
 	if err != nil {
 		return nil, err
 	}
+
+	// Release any blocks still affine to this host but no longer selected by the larger ip pool.
+	for _, block := range affBlocksToRelease {
+		blockSelectsNode := false
+		for _, pool := range allPools {
```

Comment:
What do you think about making use of this helper function here: https://github.com/projectcalico/libcalico-go/blob/26206752595de5a06705389a8302de48219016dc/lib/ipam/ipam_block_reader_writer.go#L352:29

It would want to be modified a bit so you can pass in the list of pools, to prevent having to re-query for each, but I think it would help us commonize this logic a bit.

i.e.,

```
p := c.blockReaderWriter.getPoolForIP(allPools, block)
if p != nil && !p.SelectsNode(node) {
    // Release the block affinity
}
```